platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,268 @@
+- Resetting the mutable state to the latest immutable state.
+- Restricting mutability to a single state object. Specifically, only one state can be mutable at a time.  
+- Loading snapshots from disk. States loaded from snapshots are immutable.  
+
+The interface will look something like this:
+
+```java
+public interface StateLifecycleManager {
+    /**
+     * Get the mutable state.
+     */
+    State getMutableState();
+
+   /**
+    * Reset the mutable state to the latest immutable state.
+    */
+   void resetMutableState();
+
+   /**
+    * Creates a snapshot for the latest immutable state. 
+    * 
+    * @param targetPath The path to save the snapshot.
+    */
+    State createSnapshot(final @NonNull Path targetPath);
+
+    /**
+     * Creates a mutable copy of the state. The previous mutable state becomes immutable,
+     * replacing the latest immutable state. 
+     * 
+     * @return a mutable copy of the state
+     */
+    State copyMutableState();
+}
+```
+
+#### Notes
+
+- Initial version of the interface included `getLatestImmutableState` method, but upon further consideration, 
+it was removed as it is not necessary. After the state becomes immutable, only three things can happen to it:
+  - It can be hashed
+  - It can be stored to disk as a snapshot
+  - It can become a source for a mutable copy in case of the `resetMutableState` operation
+
+  All these operations are managed by the `StateLifecycleManager` implementation, and therefore the client code does not need to access the immutable state directly.
+- The `resetMutableState` method is to effectively revert all the changes since the last copy was made. This method will require additional code changes
+  as we normally do not allow creation of immutable state copies.
+- The `createSnapshot` method requires the latest immutable state to be hashed. It will block on the `Future` returned by the `State#getHash()` method until the hash is ready.
+
+---
+
+### Implementation Details
+
+The implementation will include the methods defined in the interface and two additional static utility methods:
+
+```java
+public class StateLifecycleManagerImpl implements StateLifecycleManager {
+    /**
+     * Load an immutable copy of a state from the snapshot at the specified path.
+     *
+     * @param sourcePath the path to the snapshot
+     * @return the immutable state loaded from the snapshot
+     */
+    public static State loadSnapshot(Path sourcePath) {
+        // implementation
+    }
+
+    /**
+     * Returns a list of paths to snapshots discovered in the specified directory.
+     *
+     * @param sourcePath the path to the directory
+     * @return a list of paths to snapshots
+     */
+    public static List<Path> discover(Path sourcePath) {
+        // implementation
+    }
+}
+```
+
+
+#### Notes
+
+- The `StateLifecycleManager` implementation will maintain a **1-to-1 relationship** with the state. 
```

Comment:
I don't understand the change to being a 1-1 relationship with the state. How will it ensure a single mutable state at a given time if there is one instance per state? The app would have to keep track of the mutable state on it's own, which is one of the reasons we wanted this lifecycle manager.

Reply:
I'll clarify the wording but what I meant is that this relationship is between the lifecycle manager and a sequence of states. It is **not** a relationship between one instance of the state and one instance of the lifecycle manager. 

Imagine a hypothetical scenario in which we need to support more than one network. We will to have one lifecycle manager per network, and the lifecycle manager will have two state references at a time (current mutable state and the latest immutable state). However, these references will be updated over time and will point to other instances.